tests/end_to_end/end_to_end.cpp

```diff
@@ -828,13 +833,183 @@ map_test(ebpf_execution_type_t execution_type)
+        memcpy(return_value.local_ip6, source_ip_address.u.Byte, sizeof(return_value.local_ip6));
+        memcpy(return_value.remote_ip6, destination_ip_address.u.Byte, sizeof(return_value.remote_ip6));
+    }
+    return_value.local_port = htons(source_port);
+    return_value.remote_port = htons(destination_port);
+    return_value.protocol = ip_protocol;
+    return return_value;
+}
+
+typedef struct _ip_address
+{
+    union
+    {
+        uint32_t ipv4;
+        uint32_t ipv6[4];
+    };
+} ip_address_t;
+
+typedef struct _connection_tuple
+{
+    ip_address_t src_ip;
+    uint16_t src_port;
+    ip_address_t dst_ip;
+    uint16_t dst_port;
+    uint32_t protocol;
+} connection_tuple_t;
+
+typedef struct _connection_history
+{
+    connection_tuple_t tuple;
+    bool ipv4;
+    uint64_t start_time;
+    uint64_t end_time;
+} connection_history_t;
+
+std::string
+ip_address_to_string(bool ipv4, const ip_address_t& ip_address)
+{
+    std::string buffer;
+    if (ipv4) {
+        buffer.resize(16);
+        in_addr addr;
+        addr.S_un.S_addr = ip_address.ipv4;
+        RtlIpv4AddressToStringA(&addr, buffer.data());
+
+    } else {
+        buffer.resize(46);
+        in_addr6 addr;
+        memcpy(addr.u.Byte, ip_address.ipv6, sizeof(ip_address.ipv6));
+        RtlIpv6AddressToStringA(&addr, buffer.data());
+        buffer = "[" + buffer + "]";
+    }
+    return buffer;
+}
+
+void
+connection_tracking_test(ebpf_execution_type_t execution_type)
+{
+    _test_helper_end_to_end test_helper;
+
+    ebpf_result_t result;
+    const char* error_message = nullptr;
+    bpf_object* object = nullptr;
+    fd_t program_fd;
+    bpf_link* link;
+
+    single_instance_hook_t hook(EBPF_PROGRAM_TYPE_SOCK_OPS, EBPF_ATTACH_TYPE_CGROUP_SOCK_OPS);
+    program_info_provider_t sock_ops_program_info(EBPF_PROGRAM_TYPE_SOCK_OPS);
+
+    const char* file_name = (execution_type == EBPF_EXECUTION_NATIVE ? "conn_track_um.dll" : "conn_track.o");
+
+    result = ebpf_program_load(file_name, nullptr, nullptr, execution_type, &object, &program_fd, &error_message);
+
+    if (error_message) {
+        printf("ebpf_program_load failed with %s\n", error_message);
+        ebpf_free_string(error_message);
+        error_message = nullptr;
+    }
+
+    fd_t history_map_fd = bpf_object__find_map_fd_by_name(object, "history_map");
+    ring_buffer_opts opts{sizeof(opts)};
+    auto rb_handler = ring_buffer__new(
+        history_map_fd,
+        [](_In_ void* ctx, _In_ void* data, size_t size) -> int {
+            UNREFERENCED_PARAMETER(ctx);
+            if (size != sizeof(connection_history_t)) {
+                return 1;
+            }
+            connection_history_t* history = reinterpret_cast<connection_history_t*>(data);
+            double duration = (double)(history->end_time - history->start_time);
+            duration /= 1e9;
+            std::cout << "'" << ip_address_to_string(history->ipv4, history->tuple.src_ip) << ":"
```

Comment:
is the std:cout verified somewhere? How does the test determine if it succeeded?